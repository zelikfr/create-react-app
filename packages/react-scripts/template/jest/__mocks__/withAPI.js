import React, { Component } from 'react'

const withAPI = ComposedComponent => {
  class WithAPI extends Component {
    render () {
      return <ComposedComponent {...this.props} />
    }
  }
  return WithAPI
}

export default withAPI
