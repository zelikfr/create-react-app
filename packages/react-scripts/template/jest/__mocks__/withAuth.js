import React, { Component, createContext } from 'react'

const withAuth = () => (ComposedComponent) => {
  class WithAuth extends Component {
    render () {
      return <ComposedComponent {...this.props} />
    }
  }
  return WithAuth
}
export const AuthContext = createContext()

export default withAuth
