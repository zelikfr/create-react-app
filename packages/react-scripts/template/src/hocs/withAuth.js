import React, { Component, createContext } from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import jwtDecode from 'jwt-decode'
import { v4 } from 'uuid'
import axios from 'axios'
import config from 'config'

export const AuthContext = createContext()

const withAuth = (opts) => (ComposedComponent) => {
  const options = {
    redirectUri: '/',
    accessRestriction: 'any',
    ...opts,
  }
  class WithAuth extends Component {
    render () {
      const props = { ...this.props }
      return (
        <AuthContext.Consumer>
          {({ isAuthenticated, login, renewRefreshToken, logout }) => {
            if ((isAuthenticated && options.accessRestriction === 'notLogged') ||
              (!isAuthenticated && options.accessRestriction === 'logged')) {
              return <Redirect to={options.redirectUri} />
            }
            return <ComposedComponent {...props} login={login} logout={logout} renewRefreshToken={renewRefreshToken} />
          }}
        </AuthContext.Consumer>
      )
    }
  }

  return WithAuth
}

class AuthProviderComponent extends Component {
  constructor (props) {
    super(props)

    const refreshToken = window.localStorage.getItem('refreshToken')
    const expirationDate = window.localStorage.getItem('expirationDate')
    const userId = window.localStorage.getItem('userId')

    this.request = {}
    this.loginRequest = {}

    this.state = {
      isAuthenticated: !!refreshToken,
      refreshToken,
      expirationDate,
      userId,
    }
  }
  componentDidMount () {
    if (this.state.refreshToken) {
      this.renewRefreshToken()
    }
  }
  componentWillUnmount () {
    this._isMounted = false
    if (this.request.token) {
      this.request.cancel()
    }
    if (this.loginRequest.token) {
      this.loginRequest.cancel()
    }
  }
  authenticateUser (refreshToken) {
    window.localStorage.setItem('refreshToken', refreshToken)
    const { exp: expirationDate, global_user_id: userId } = jwtDecode(refreshToken)
    window.localStorage.setItem('refreshTokenExpirationDate', expirationDate)
    window.localStorage.setItem('userId', userId)
    this.setState({ isAuthenticated: true, expirationDate, userId })
  }
  logout () {
    window.localStorage.removeItem('refreshToken')
    window.localStorage.removeItem('refreshTokenExpirationDate')
    window.localStorage.removeItem('userId')
    this.setState({
      isAuthenticated: false,
      refreshToken: null,
      expirationDate: null,
      userId: null,
    })
  }
  async login (email, password) {
    if (this.loginRequest.token) {
      this.loginRequest.cancel()
    }
    this.loginRequest = axios.CancelToken.source()
    try {
      const { data: res } = await axios.request(
        {
          url: config.authUrl + '/api.companyusermasterdata.default/command',
          cancelToken: this.loginRequest.token,
          method: 'POST',
          data: {
            '$type': 'TraceOne.Api.CompanyUserMasterData.Default.Messages.Commands.AuthenticateUserCommand, TraceOne.Api.CompanyUserMasterData.Default.Messages',
            'messageId': v4(),
            email,
            password,
          },
        })

      if (res.succeeded) {
        this.authenticateUser(res.refreshToken)
        return true
      }
    } catch (err) {
      return err.error
    }
  }
  shouldRefreshToken (expirationDateToken) {
    return (expirationDateToken - Date.now() / 1000 < 7 * 60)
  }
  async renewRefreshToken () {
    if (this.shouldRefreshToken(this.state.expirationDate)) {
      if (this.request.token) {
        this.request.cancel()
      }
      this.request = axios.CancelToken.source()
      try {
        const { data: res } = await axios.request({
          url: config.authUrl + '/api.companyusermasterdata.default/command',
          method: 'POST',
          cancelToken: this.request.token,
          data: {
            '$type': 'TraceOne.Api.CompanyUserMasterData.Default.Messages.Commands.RenewRefreshTokenCommand, TraceOne.Api.CompanyUserMasterData.Default.Messages',
            'messageId': v4(),
            'RefreshToken': `${this.state.refreshToken}`,
            'UserId': `${this.state.userId}`,
          },
        })

        if (res.succeeded) {
          this.authenticateUser(res.refreshToken)
          return true
        }
      } catch (err) {
        if (!axios.isCancel(err)) {
          this.logout()
        }
      }
    }
    return false
  }
  render () {
    const { isAuthenticated } = this.state
    return (
      <AuthContext.Provider
        value={{
          isAuthenticated,
          login: (email, password) => this.login(email, password),
          renewRefreshToken: () => this.renewRefreshToken(),
          logout: () => this.logout(),
        }}
      >
        {this.props.children}
      </AuthContext.Provider>
    )
  }
}

AuthProviderComponent.propTypes = {
  children: PropTypes.any,
}

export const AuthProvider = (AuthProviderComponent)

export default withAuth
