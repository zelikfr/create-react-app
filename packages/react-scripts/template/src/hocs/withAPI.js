import React, { Component } from 'react'
import axios from 'axios'
import config from 'config'
import { AuthContext } from 'hocs/withAuth'

const withAPI = ComposedComponent => {
  class WithAPI extends Component {
    constructor (props) {
      super(props)
      this.getOptions = this.getOptions.bind(this)
      this.call = this.call.bind(this)
      this.multipleCalls = this.multipleCalls.bind(this)
      const CancelToken = axios.CancelToken
      this.request = CancelToken.source()
    }
    componentWillUnmount () {
      this.request.cancel()
    }
    getOptions (opts) {
      const options = {
        ...opts,
        method: opts.method,
        url: `${opts.apiUrl || config.apiUrl}/${opts.url}`,
        cancelToken: this.request.token,
      }
      return options
    }
    handleError (err, error, prevQuery) {
      if (!axios.isCancel(err) && error) {
        error(err)
      }
    }
    multipleCalls (queries, success, error) {
      const queriesOpts = queries.map(q => axios(this.getOptions(q)))
      const customSpread = function (...args) {
        success(...[...args].map(d => d.data))
      }
      const refreshToken = this.context.renewRefreshToken()
      if (refreshToken) {
        axios.all(queriesOpts)
          .then(axios.spread(customSpread))
          .catch((err) => this.handleError(err, error, { type: 'multipleCalls', queries, success, error }))
      }
    }
    call (query, success, error) {
      const refreshToken = this.context.renewRefreshToken()
      if (refreshToken) {
        axios.request(this.getOptions(query))
          .then((res) => success(res.data))
          .catch((err) => this.handleError(err, error, { type: 'call', query, success, error }))
      }
    }
    render () {
      return <ComposedComponent {...this.props} call={this.call} multipleCalls={this.multipleCalls} />
    }
  }

  WithAPI.contextType = AuthContext

  return WithAPI
}

export default withAPI
