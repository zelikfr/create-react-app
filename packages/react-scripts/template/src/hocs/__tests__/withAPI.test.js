import React from 'react'
import { mount } from 'enzyme'
import withAPI from '../withAPI.js'
import axios from 'axios'
import { AuthContext } from 'hocs/withAuth'

const Component = () => (
  <div></div>
)

describe('withAPI', () => {
  it('should create a withAPI instance with props', () => {
    // When I render the withAPI element
    const HocComponent = withAPI(Component)
    const component = mount(<HocComponent />)
    // Then I should have the component to match the snapshot.
    expect(component).toMatchSnapshot()
  })
  it('should unmount a withAPI instance', () => {
    // When I render the withAPI element
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const spyCancel = spyOn(component.instance().request, 'cancel')
    component.unmount()
    // Then I should have the component to match the snapshot.
    expect(spyCancel).toHaveBeenCalled()
  })
  it('should success axios call', done => {
    // When I render the withAPI element and do a call
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const err = jest.fn()
    const promise = Promise.resolve({ data: 'myData' })
    const spyCreate = spyOn(axios, 'request').and.callFake(() => promise)
    component.instance().call({ method: 'GET', url: 'api/endpoint' }, succ, err)
    // Then spyCreate is called and success callback is called too
    expect(spyCreate).toHaveBeenCalled()
    promise.then(() => {
      expect(succ).toHaveBeenCalledWith('myData')
      done()
    })
  })
  it('should fail axios call', done => {
    // When I render the withAPI element and fail a call
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const err = jest.fn()
    const rejectValue = { error: 'error' }
    const promise = Promise.reject(rejectValue)
    const handleError = spyOn(component.instance(), 'handleError')
    const spyCreate = spyOn(axios, 'request').and.callFake(() => promise)
    component.instance().call({ method: 'GET', url: 'api/endpoint' }, succ, err)
    expect(spyCreate).toHaveBeenCalled()
    // Then spyCreate is called and error callback is called too
    promise.then(() => { }).catch(() => {
      expect(handleError).toHaveBeenCalled()
      done()
    })
  })
  it('should success axios multiplecall', done => {
    // When I render the withAPI element and do a multiple call
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const err = jest.fn()
    const promise = Promise.resolve({ data: 'myData' })
    const spyCreate = spyOn(axios, 'all').and.callFake(() => promise)
    component.instance().multipleCalls([{ method: 'GET', url: 'api/endpoint' }, { method: 'GET', url: 'api/endpoint2' }], succ, err)
    // Then spyCreate is called and success callback is called too
    expect(spyCreate).toHaveBeenCalled()
    promise.then(() => {
      expect(succ).toHaveBeenCalled()
      done()
    })
  })
  it('should call axios multiplecall with Expired token', done => {
    // When I render the withAPI element and do a multiple call
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => false }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const err = jest.fn()
    const promise = Promise.resolve({ data: 'myData' })
    const spyCreate = spyOn(axios, 'all').and.callFake(() => promise)
    component.instance().multipleCalls([{ method: 'GET', url: 'api/endpoint' }, { method: 'GET', url: 'api/endpoint2' }], succ, err)
    // Then spyCreate is called and success callback is called too
    expect(spyCreate).not.toHaveBeenCalled()
    promise.then(() => {
      expect(succ).not.toHaveBeenCalled()
      done()
    })
  })
  it('should fail axios multiplecall', done => {
    // When I render the withAPI element and fail a multiple calls
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const err = jest.fn()
    const rejectValue = { error: 'error' }
    const promise = Promise.reject(rejectValue)
    const handleError = spyOn(component.instance(), 'handleError')
    const spyCreate = spyOn(axios, 'all').and.callFake(() => promise)
    component.instance().multipleCalls([{ method: 'GET', url: 'api/endpoint' }, { method: 'GET', url: 'api/endpoint2' }], succ, err)
    // Then spyCreate is called and error callback is called too
    expect(spyCreate).toHaveBeenCalled()
    promise.then(() => { }).catch(() => {
      expect(handleError).toHaveBeenCalled()
      done()
    })
  })
  it('should fail axios call without error cb', done => {
    // When I render the withAPI element and fail a call
    const HocComponent = withAPI(Component)
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const succ = jest.fn()
    const rejectValue = { error: 'invalid_token' }
    const promise = Promise.reject(rejectValue)
    const handleError = spyOn(component.instance(), 'handleError').and.callThrough()
    const spyCreate = spyOn(axios, 'request').and.callFake(() => promise)
    component.instance().call({ method: 'GET', url: 'api/endpoint' }, succ)
    // Then spyCreate is called and error callback is called too
    expect(spyCreate).toHaveBeenCalled()
    promise.then(() => { }).catch(() => {
      expect(handleError).toHaveBeenCalled()
      done()
    })
  })
  it('should test handleError method no new access_token provided', done => {
    // When I render the withAPI element
    const HocComponent = withAPI(Component)
    const login = jest.fn()
    const component = mount(<AuthContext.Provider value={{ isAuthenticated: false, login: () => jest.fn(), logout: () => jest.fn(), renewRefreshToken: () => jest.fn() }}><HocComponent /></AuthContext.Provider>, { selector: 'WithAPI' })
    const resError = { response: { data: { error: 'invalid_token' } } }
    const resSucc = { data: {} }
    const promise = Promise.resolve(resSucc)
    spyOn(axios, 'request').and.callFake(() => promise)
    const err = jest.fn()
    const spyCall = spyOn(component.instance(), 'call')
    const prevQuery = { query: { method: 'GET', url: 'api/endpoint' }, type: 'call' }
    component.setProps({ login })
    component.instance().handleError(resError, err, prevQuery)
    // Then login and spyCall are not called
    promise.then(() => {
      expect(login).not.toHaveBeenCalled()
      expect(spyCall).not.toHaveBeenCalled()
      done()
    })
  })
})
