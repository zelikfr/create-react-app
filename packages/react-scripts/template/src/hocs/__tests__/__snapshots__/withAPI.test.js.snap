// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withAPI should create a withAPI instance with props 1`] = `
<WithAPI>
  <Component
    call={[Function]}
    multipleCalls={[Function]}
  >
    <div />
  </Component>
</WithAPI>
`;
